### PR TITLE
Celery schedules fix

### PR DESCRIPTION
**Pull request by shawnjoseph**
**Summary:** merge in fix for GC-6953 to fix schedule execution error as well as an empty heap at startup error

* fixes cherry-picked from celery/django-celery-beat to avoid merging in any breaking code

* https://github.com/celery/django-celery-beat/commit/ab2b5be36b43eaf9b5d8f72e05d23e6a8830ba5e
* https://github.com/celery/django-celery-beat/commit/048573955fc376ab8ec093ef36da254d2b687e3c

### CODE REVIEW

**Review by ramonPaulAlvarez — APPROVED** [review]

File: t/unit/test_schedulers.py [hunk 1]
```diff
@@ -400,6 +400,28 @@ def test_sync_rollback_on_save_error(self):
...
+        # heap size is constant unless the schedule changes
+        monkeypatch.setattr(self.s, 'schedule_changed', lambda: True)
+        expected_heap_size = len(self.s.schedule.values())
+        self.s.tick()
+        assert len(self.s._heap) == expected_heap_size
+        self.s.tick()
+        assert len(self.s._heap) == expected_heap_size
+
+    def test_scheduler_schedules_equality_on_change(self, monkeypatch):
+        monkeypatch.setattr(self.s, 'schedule_changed', lambda: False)
```
Comment (LourdesMoran) [inline t/unit/test_schedulers.py:419]: Out of curiosity, how does this functions gets the value of `monkeypatch`?

File: t/unit/test_schedulers.py [hunk 2]
```diff
@@ -400,6 +400,28 @@ def test_sync_rollback_on_save_error(self):
...
+        assert self.s.schedules_equal(self.s.schedule, self.s.schedule)
+
+        monkeypatch.setattr(self.s, 'schedule_changed', lambda: True)
+        assert not self.s.schedules_equal(self.s.schedule, self.s.schedule)
```
Comment (LourdesMoran) [inline t/unit/test_schedulers.py:423]: hmm, this is probably ok, but how is the same object not equal here?

File: t/unit/test_schedulers.py (hunk 1 above)
Comment (shawnjoseph) [reply t/unit/test_schedulers.py:419]: `monkeypatch` is a pytest plugin. By adding `monkeypatch` to the test function signature pytest provides the `monkeypatch` object.

File: t/unit/test_schedulers.py (hunk 2 above)
Comment (shawnjoseph) [reply t/unit/test_schedulers.py:423]: hmm, I'm not really sure

File: t/unit/test_schedulers.py (hunk 1 above)
Comment (LourdesMoran) [reply t/unit/test_schedulers.py:419]: hmm, but how does the monkeypatch knows what to patch?

File: t/unit/test_schedulers.py (hunk 2 above)
Comment (LourdesMoran) [reply t/unit/test_schedulers.py:423]: how do you know the test is ok then?

**Pull request merged**